blog-posts/event-sourcing-serveless/event-sourcing-serveless-plan.md

```diff
@@ -0,0 +1,105 @@
+---
+published: false
+title: 'The best way to do event sourcing in serverless'
+cover_image:
+description: 'Description of the article'
+tags: serverless, event-sourcing, AWS, CQRS
+series:
+canonical_url:
+---
+
+# Introduction
+
+You probably use a bank application in your every day life and you wonder how they work. They rely on event sourcing to store their data and they compute the balance of your account from all credit/debit events stored. Numerous applications need to store a trace of all events to keep and use the history. If you have that need and you're thinking about implementing it in serverless, you are in the right place. Thus I want to share with you, 2 years of learnings that will lead you to have the most stable and the less painful solution you can go with.
+
+## CQRS and event sourcing
+
+CQRS (Command and Query Responsibility Segregation) is a pattern that separates read and write operations for a data store. It's a good thing to grasp a bit what it is to understand better what will be approached in this article, but don't worry it is not required as it is quite straightforward. Still numerous resources already deal with this topic that you can easily look up on the web.
+
+Event sourcing is a way to implement the data store in a CQRS pattern interface. There are also a lot of resources on the internet about it if you want to quickly take a look at it before deep diving into our banking application.
+
+## What we will build together
+
+Let's build the example of the banking application. We want to store data with the event sourcing pattern to be able to create any data view models from the list of credit/debit events of a user. For reading and writing data, we want to set up a CQRS interface which fits perfectly with the event sourcing data storing system. For now Let's take a look at what the architecture of such an application should look like.
+
+_insert schema of archi agnostic of technologies_
+
+We have two projections, one to get the total balance of your account and one to get the current month balance. The synchronization of the read models with the event table which is the source of truth relies on an Event Driven Architecture. We react to writing events in the event table to trigger updates of the read models.
+
+Now let's see how to implement it with AWS serverless technologies.
+
+# Plan
+
+- **EVENT LEDGER DB**
+  - source of truth of your application data in which you store every event that occurs in your account.
+  - Use DynamoDB to store credit/debit events
+    - pk=idUser, sk='CreditEvent'|'DebitEvent', amount=number, version=number
```

Comment:
pretty sure `sk` is the version number in most event sourcing we did in kumo, not event type